DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs

```diff
@@ -463,5 +635,47 @@ private void AddProgressRecordForGroupWithSharedCourse()
                     SET IDENTITY_INSERT dbo.Progress OFF"
             );
         }
+
+        private void AddDelegateGroupCustomisationIfNotExist(int groupId, int customisationId)

```

Comment:
Ideally, we want to minimise the number of methods defined purely for tests that affect the DB. If we need to test a certain functionality, then in most cases there should already be a record(s) that exists that matches the criteria we need to test and thus that record(s) should be used. 

Though in some cases there may be a need to add new data or update existing records if none of it quite fits. In which case, we should avoid defining methods like this, and instead follow the procedure listed in the documentation (at this point I would link the documentation page, but the page's creation is an action item for Steve from the last retro and it looks like he hasn't done it yet, I'll chase him up on it when he is back tomorrow). 

Off the top of my head, this procedure involves creating a new SQL script that does all the insertions/updates necessary and plonking it in the SQLScripts folder. Full details should be in the documentation page (so it might be worth waiting for Steve to get that up?).

Reply:
Actually, I had the wrong idea. With the new docs we do prefer setting up the data temporarily inside a test for most cases, unless data is reused multiple times. Docs for reference: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/19880509461/Adding+test+data
So feel free to keep the methods and their use (apologies for any confusion). I will have another look over this GroupsDataServiceTests file soon, but I think the rest of my comments still stand.

Reply:
I've had another look over this now. I think these tests are suitable and don't need much modification, mostly just minor cleaning. I've resolved any comments that I think no longer apply. Apologies for the delay and any confusion.

Reply:
So in a previous PR I expressed a concern about relaince on a static database. I think the follow up to that is one that we should have enough confidence that this test database will remain static and we can write tests with that in mind.

So with that in mind I've decided to refactor these tests anyway and I think the end result is actually much better.
This has lead to the removal of all of those SQL helper functions to create data.

I spent a bit of time going through the test database and have managed to locate almost enough data to cover the various test cases.
There was just one case the test db did not hold but luckily there is a SQL helper method that creates the exact data I needed so have utlised that: `AddDelegateToGroupWithSharedCourse()`.

I think your initial review comments still stand really despite the, now, preference for inline test data creation. Because this has lead to a good refactor.

Reply:
Yep, I'm liking these changes. They are consistent with other tests and only create data when necessary (hence the other method that is being reused here).